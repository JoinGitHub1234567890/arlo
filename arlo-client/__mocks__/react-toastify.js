/**
 * Mock for react-toastify
 */

const toast = jest.genMockFromModule('react-toastify')

toast.error = jest.fn()

export default { toast }
