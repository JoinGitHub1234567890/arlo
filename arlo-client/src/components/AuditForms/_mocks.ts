import { Audit } from '../../types'

export const mockAudit: Audit = {
  name: 'Primary 2019',
  riskLimit: 10,
  randomSeed: 123123123,

  contests: [
    {
      id: 'contest-1',
      name: 'Contest 1',

      choices: [
        {
          id: 'candidate-1',
          name: 'Candidate 1',
          numVotes: 42,
        },
        {
          id: 'candidate-2',
          name: 'Candidate 2',
          numVotes: 72,
        },
      ],

      totalBallotsCast: 4200,
    },
  ],

  jurisdictions: [
    {
      id: 'adams-county',
      name: 'Adams County',
      contests: ['contest-1'],
      auditBoards: [
        {
          id: 'audit-board-1',
          members: [],
        },
        {
          id: 'audit-board-2',
          members: [],
        },
      ],
      ballotManifest: {
        filename: 'Adams_County_Manifest.csv',
        numBallots: 123456,
        numBatches: 560,
        uploadedAt: '2019-06-17 11:45:00',
      },
    },
  ],

  rounds: [
    {
      startedAt: '2019-06-17 11:45:00',
      endedAt: '2019-06-17 11:55:00',
      contests: [
        {
          id: 'contest-1',
          endMeasurements: {
            pvalue: 0.085,
            isComplete: false,
          },
          results: {
            'candidate-1': 55,
            'candidate-2': 35,
          },
          sampleSize: 25,
        },
      ],
      jurisdictions: {
        'adams-county': {
          numBallots: 15,
        },
      },
    },
  ],
}

export default mockAudit
