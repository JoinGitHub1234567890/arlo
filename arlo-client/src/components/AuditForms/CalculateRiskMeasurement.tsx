import React from 'react'
import styled from 'styled-components'
import { toast } from 'react-toastify'
import { Formik, FormikProps, FieldArray, Form, Field } from 'formik'
// import * as Yup from 'yup'
import FormSection, {
  FormSectionLabel,
  FormSectionDescription,
} from '../Form/FormSection'
import FormWrapper from '../Form/FormWrapper'
import FormButton from '../Form/FormButton'
import FormField from '../Form/FormField'
import FormButtonBar from '../Form/FormButtonBar'
import { api } from '../utilities'
import { Contest, Round, Candidate, RoundContest } from '../../types'

const InputSection = styled.div`
  display: block;
  margin-top: 25px;
  width: 100%;
  font-size: 0.4em;
`

const InputLabel = styled.label`
  display: inline-block;
`

const InlineInput = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: 10px;
  width: 50%;
`

interface Props {
  audit: any
  isLoading: boolean
  setIsLoading: (isLoading: boolean) => void
  updateAudit: () => void
}

interface CalculateRiskMeasurementValues {
  round: number
  contests: {
    [key: string]: number | ''
  }[]
}

type AggregateContest = Contest & RoundContest

const CalculateRiskMeasurmeent = (props: Props) => {
  const { audit, isLoading, setIsLoading, updateAudit } = props

  //const sumOfAuditedVotes: { current: number } = useRef(0)

  const downloadBallotRetrievalList = (id: number, e: any) => {
    e.preventDefault()
    const jurisdictionID: string = audit.jurisdictions[0].id
    window.open(`/jurisdiction/${jurisdictionID}/${id}/retrieval-list`)
  }

  const downloadAuditReport = async (e: React.MouseEvent) => {
    e.preventDefault()
    try {
      window.open(`/audit/report`)
      updateAudit()
    } catch (err) {
      toast.error(err.message)
    }
  }

  const calculateRiskMeasurement = async (
    values: CalculateRiskMeasurementValues
  ) => {
    try {
      const jurisdictionID: string = audit.jurisdictions[0].id
      const body: any = {
        contests: audit.contests.map((contest: Contest, i: number) => ({
          // map values.contests
          id: contest.id,
          results: {
            ...values.contests[i],
          },
        })),
      }

      setIsLoading(true)
      await api(`/jurisdiction/${jurisdictionID}/${values.round}/results`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      })
      //sumOfAuditedVotes.current +=
      //  Number(values['candidate-1']) + Number(values['candidate-2'])
      updateAudit()
    } catch (err) {
      toast.error(err.message)
    }
  }

  const rounds: CalculateRiskMeasurementValues[] = audit.rounds.map(
    (r: Round, i: number) => ({
      contests: audit.contests.map((contest: Contest, j: number) => ({
        ...contest.choices.reduce((acc, choice: Candidate) => {
          return { ...acc, [choice.id]: r.contests[j].results[choice.id] || '' }
        }, {}),
      })),
      round: i + 1,
    })
  )

  return audit.rounds.map((round: Round, i: number) => {
    const aggregateContests: AggregateContest[] = audit.contests.reduce(
      (acc: any[], contest: Contest) => {
        const roundContest = round.contests.find(v => v.id === contest.id)
        acc.push({ ...contest, ...roundContest })
        return acc
      },
      []
    )
    const completeContests = aggregateContests.reduce((acc, c) => {
      if (c.endMeasurements.isComplete) {
        acc += 1
      }
      return acc
    }, 0)
    /*
      const overCount = function(this: {
      parent: { [key: string]: number }
    }): boolean {
      const count = Object.values(this.parent).reduce(
        (c: number, v: number) => {
          c += v || 0
          return c
        },
        0
      )
      const maxVotes = this.parent.parent.id
      return count + sumOfAuditedVotes.current <= max
    }
    const schema = Yup.object().shape({
      contests: Yup.lazy(() => {
        return Object.keys(v.contests).reduce((acc: any, result: string) => {
          acc[result] = Yup.number().test(
            'overCount',
            'Cannot exceed the number of total ballots cast',
            overCount
          )
          return acc
        }, {})
      })
    }) */
    /* eslint-disable react/no-array-index-key */
    const aggregatedBallots = aggregateContests.reduce(
      (acc: number, contest: AggregateContest) => {
        acc += contest.sampleSize
        return acc
      },
      0
    )
    return (
      <Formik
        key={i}
        onSubmit={calculateRiskMeasurement}
        initialValues={rounds[i]}
        //validationSchema={schema}
        enableReinitialize
        render={({
          values,
          handleSubmit,
        }: FormikProps<CalculateRiskMeasurementValues>) => (
          <Form>
            <FormWrapper title={`Round ${i + 1}`}>
              <FormSectionLabel>
                Ballot Retrieval List: {aggregatedBallots} Total Ballots
              </FormSectionLabel>
              <FormSectionDescription>
                {aggregateContests.map(
                  (contest: AggregateContest, i: number) => (
                    <p key={contest.id}>
                      Contest {i + 1}: {contest.sampleSize} ballots
                    </p>
                  )
                )}
              </FormSectionDescription>
              {/*<SectionLabel>
                Ballot Retrieval List \n
                {contest ? `${contest.sampleSize} Ballots` : ''}
              </SectionLabel>*/}
              <FormButton
                onClick={(e: React.MouseEvent) =>
                  downloadBallotRetrievalList(i + 1, e)
                }
                inline
              >
                Download Aggregated Ballot Retrieval List for Round {i + 1}
              </FormButton>
              {/** contest iteration */}
              <FieldArray
                name="contests"
                render={() => {
                  return (
                    <>
                      {values.contests.map((contest, j) => {
                        return (
                          <FormSection
                            key={aggregateContests[j].id}
                            label={`Contest ${j + 1}: ${
                              aggregateContests[j].name
                            }`}
                          >
                            {aggregateContests[j].endMeasurements.isComplete ? (
                              <>
                                <FormSectionLabel>
                                  Contest Status: COMPLETE
                                </FormSectionLabel>
                                <InputSection>
                                  <InlineInput>
                                    <InputLabel>Risk Limit: </InputLabel>
                                    {audit.riskLimit}%
                                  </InlineInput>
                                  <InlineInput>
                                    <InputLabel>P-value: </InputLabel>{' '}
                                    {
                                      aggregateContests[j].endMeasurements
                                        .pvalue
                                    }
                                  </InlineInput>
                                </InputSection>
                              </>
                            ) : (
                              <>
                                <FormSectionLabel>
                                  Audited Results: Round {i + 1}, Contest{' '}
                                  {j + 1} INCOMPLETE
                                </FormSectionLabel>
                                <FormSectionDescription>
                                  Enter the number of votes recorded for each
                                  candidate/choice in the audited ballots for
                                  Round {i + 1}, Contest {j + 1}
                                </FormSectionDescription>
                                <InputSection>
                                  {Object.keys(contest).map(choiceId => {
                                    const name = aggregateContests[
                                      j
                                    ].choices.find(
                                      (candidate: Candidate) =>
                                        candidate.id === choiceId
                                    )!.name
                                    return (
                                      <React.Fragment key={choiceId}>
                                        <InlineInput>
                                          <InputLabel>{name}</InputLabel>
                                          <Field
                                            name={`contests[${j}][${choiceId}]`}
                                            type="number"
                                            component={FormField}
                                          />
                                        </InlineInput>
                                      </React.Fragment>
                                    )
                                  })}
                                </InputSection>
                              </>
                            )}
                          </FormSection>
                        )
                      })}
                    </>
                  )
                }}
              />
              {isLoading && <p>Loading...</p>}
              <FormSection>
                <FormSectionLabel>
                  Audit Progress: {completeContests} of {audit.contests.length}{' '}
                  complete
                </FormSectionLabel>
              </FormSection>
              {i + 1 === audit.rounds.length &&
                aggregateContests.some(
                  (contest: AggregateContest) =>
                    !contest.endMeasurements.isComplete
                ) &&
                !isLoading && (
                  <FormButtonBar>
                    <FormButton type="button" onClick={handleSubmit}>
                      Calculate Risk Measurement
                    </FormButton>
                  </FormButtonBar>
                )}
              <FormSection>
                {/*<FormSectionLabel>
                  Audit Status:{' '}
                  {contest.endMeasurements.isComplete
                    ? 'COMPLETE'
                    : 'INCOMPLETE'}
                </FormSectionLabel>
                <InputSection>
                  <InlineInput>
                    <InputLabel>Risk Limit: </InputLabel>
                    {audit.riskLimit}%
                  </InlineInput>
                  <InlineInput>
                    <InputLabel>P-value: </InputLabel>{' '}
                    {contest.endMeasurements.pvalue}
                  </InlineInput>
                </InputSection>*/}
                {/* {Form 3} */}
                {completeContests === audit.contests.length && (
                  <FormButton
                    onClick={(e: React.MouseEvent) => downloadAuditReport(e)}
                    size="sm"
                    inline
                  >
                    Download Audit Report
                  </FormButton>
                )}
              </FormSection>
            </FormWrapper>
          </Form>
        )}
      />
    )
  })
}

export default React.memo(CalculateRiskMeasurmeent)
