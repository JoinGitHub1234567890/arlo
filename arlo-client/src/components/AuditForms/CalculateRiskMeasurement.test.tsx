import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { toast } from 'react-toastify'
import CalculateRiskMeasurement from './CalculateRiskMeasurement'
import { statusStates } from './_mocks'
import apiMock from '../utilities'

jest.mock('../utilities')
//jest.mock('react-toastify')

const setIsLoadingMock = jest.fn()
const updateAuditMock = jest.fn()

beforeEach(() => {
  ;(setIsLoadingMock as jest.Mock).mockReset()
  ;(updateAuditMock as jest.Mock).mockReset()
})

describe('CalculateRiskMeasurement', () => {
  it('renders first round correctly', () => {
    const container = render(
      <CalculateRiskMeasurement
        audit={statusStates[3]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders completion in first round correctly', () => {
    const container = render(
      <CalculateRiskMeasurement
        audit={statusStates[4]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it(`handles inputs`, async () => {
    const toastSpy = jest.spyOn(toast, 'error').mockImplementation()
    ;(apiMock as jest.Mock).mockImplementation(() =>
      Promise.resolve({
        message: 'success',
        ok: true,
      })
    )
    const { container, getByTestId, queryAllByText, getByText } = render(
      <CalculateRiskMeasurement
        audit={statusStates[3]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
      />
    )

    expect(container).toMatchSnapshot()

    const choiceOne: any = getByTestId(`round-0-contest-0-choice-choice-1`)
    const choiceTwo: any = getByTestId(`round-0-contest-0-choice-choice-2`)

    fireEvent.change(choiceOne, { target: { value: -1 } })
    fireEvent.change(choiceTwo, { target: { value: -1 } })
    fireEvent.blur(choiceOne)
    fireEvent.blur(choiceTwo)
    expect(choiceOne.value).toBe('-1')
    expect(choiceTwo.value).toBe('-1')
    await wait(() => {
      expect(queryAllByText('Must be a positive number').length).toBe(2)
    })

    fireEvent.change(choiceOne, { target: { value: '0.5' } })
    fireEvent.change(choiceTwo, { target: { value: '0.5' } })
    fireEvent.blur(choiceOne)
    fireEvent.blur(choiceTwo)
    expect(choiceOne.value).toBe('0.5')
    expect(choiceTwo.value).toBe('0.5')
    await wait(() => {
      expect(queryAllByText('Must be an integer').length).toBe(2)
    })

    fireEvent.change(choiceOne, { target: { value: '' } })
    fireEvent.change(choiceTwo, { target: { value: '' } })
    fireEvent.blur(choiceOne)
    fireEvent.blur(choiceTwo)
    expect(choiceOne.value).toBe('')
    expect(choiceTwo.value).toBe('')
    await wait(() => {
      expect(queryAllByText('Must be a number').length).toBe(2)
    })

    fireEvent.change(choiceOne, { target: { value: '5' } })
    fireEvent.change(choiceTwo, { target: { value: '5' } })
    fireEvent.blur(choiceOne)
    fireEvent.blur(choiceTwo)
    expect(choiceOne.value).toBe('5')
    expect(choiceTwo.value).toBe('5')
    fireEvent.click(getByText('Calculate Risk Measurement'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(setIsLoadingMock).toBeCalledTimes(1)
      expect(updateAuditMock).toBeCalledTimes(1)
      expect(toastSpy).toBeCalledTimes(0)
    })
  })

  it('downloads aggregated ballots report', () => {
    ;(global as any).open = jest.fn()
    const { getByText } = render(
      <CalculateRiskMeasurement
        audit={statusStates[3]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
      />
    )

    fireEvent.click(
      getByText('Download Aggregated Ballot Retrieval List for Round 1'),
      { bubbles: true }
    )

    expect((global as any).open).toHaveBeenCalledTimes(1)
    expect((global as any).open).toHaveBeenCalledWith(
      `/jurisdiction/jurisdiction-1/1/retrieval-list`
    )
  })

  it('downloads audit report', () => {
    ;(global as any).open = jest.fn()
    const { getByText } = render(
      <CalculateRiskMeasurement
        audit={statusStates[4]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
      />
    )

    fireEvent.click(getByText('Download Audit Report'), { bubbles: true })

    expect((global as any).open).toHaveBeenCalledTimes(1)
    expect((global as any).open).toHaveBeenCalledWith(`/audit/report`)
  })

  it('handles errors from api', async () => {
    ;(apiMock as jest.Mock).mockReset()
    ;(apiMock as jest.Mock).mockImplementation(() =>
      Promise.reject({
        message: 'error',
        ok: false,
      })
    )
    const toastSpy = jest.spyOn(toast, 'error').mockImplementation()
    const { getByTestId, getByText } = render(
      <CalculateRiskMeasurement
        audit={statusStates[3]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
      />
    )

    const choiceOne: any = getByTestId(`round-0-contest-0-choice-choice-1`)
    const choiceTwo: any = getByTestId(`round-0-contest-0-choice-choice-2`)

    fireEvent.change(choiceOne, { target: { value: '5' } })
    fireEvent.change(choiceTwo, { target: { value: '5' } })
    fireEvent.click(getByText('Calculate Risk Measurement'), { bubbles: true })

    await wait(() => {
      expect(apiMock).toBeCalledTimes(1)
      expect(setIsLoadingMock).toBeCalledTimes(1)
      expect(updateAuditMock).toBeCalledTimes(0)
      expect(toastSpy).toBeCalledTimes(1)
    })
  })
})
