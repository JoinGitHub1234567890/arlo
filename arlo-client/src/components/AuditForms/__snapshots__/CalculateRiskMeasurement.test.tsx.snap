// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculateRiskMeasurement handles inputs 1`] = `
<div>
  <form>
    <div
      class="sc-ifAKCX jXnARC"
    >
      <div
        class="sc-bxivhb doxbWk"
      >
        Round 1
      </div>
      <div
        class="sc-htpNat gYCOid"
      >
        Ballot Retrieval List: 
        379
         Total Ballots
      </div>
      <div
        class="sc-bwzfXH xctbO"
      >
        <p>
          Contest 
          1
          : 
          379
           ballots
        </p>
      </div>
      <button
        class="sc-bZQynM cjVqOQ"
      >
        Download Aggregated Ballot Retrieval List for Round 
        1
      </button>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <div
          class="sc-htpNat gYCOid"
        >
          Contest 1: contest name
        </div>
        <div
          class="sc-htpNat gYCOid"
        >
          Audited Results: Round 
          1
          , Contest 
          1
           
          INCOMPLETE
        </div>
        <div
          class="sc-bwzfXH xctbO"
        >
          Enter the number of votes recorded for each candidate/choice in the audited ballots for Round 
          1
          , Contest 
          1
        </div>
        <div
          class="sc-gqjmRU bqmubQ"
        >
          <div
            class="sc-jTzLTM hJXymL"
          >
            <label
              class="sc-VigVT hLPYRt"
            >
              choice one
            </label>
            <div
              class="sc-htoDjs jjofgS"
            >
              <input
                class="sc-dnqmqq dvJRiU"
                data-testid="round-0-contest-0-choice-choice-1"
                name="contests[0][choice-1]"
                type="number"
                value="0"
              />
            </div>
          </div>
          <div
            class="sc-jTzLTM hJXymL"
          >
            <label
              class="sc-VigVT hLPYRt"
            >
              choice two
            </label>
            <div
              class="sc-htoDjs jjofgS"
            >
              <input
                class="sc-dnqmqq dvJRiU"
                data-testid="round-0-contest-0-choice-choice-2"
                name="contests[0][choice-2]"
                type="number"
                value="0"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-bdVaJa fdbLH"
      >
        <div
          class="sc-htpNat gYCOid"
        >
          Audit Progress: 
          0
           of 
          1
           
          complete
        </div>
      </div>
      <div
        class="sc-gZMcBi DmELe"
      >
        <button
          class="sc-EHOje iYzqVS"
          type="button"
        >
          Calculate Risk Measurement
        </button>
      </div>
      <div
        class="sc-bdVaJa fdbLH"
      />
    </div>
  </form>
</div>
`;

exports[`CalculateRiskMeasurement renders completion in first round correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="sc-ifAKCX jXnARC"
        >
          <div
            class="sc-bxivhb doxbWk"
          >
            Round 1
          </div>
          <div
            class="sc-htpNat gYCOid"
          >
            Ballot Retrieval List: 
            379
             Total Ballots
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            <p>
              Contest 
              1
              : 
              379
               ballots
            </p>
          </div>
          <button
            class="sc-bZQynM cjVqOQ"
          >
            Download Aggregated Ballot Retrieval List for Round 
            1
          </button>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Contest 1: contest name
            </div>
            <div
              class="sc-htpNat gYCOid"
            >
              Audited Results: Round 
              1
              , Contest 
              1
               
              COMPLETE
            </div>
            <div
              class="sc-gqjmRU bqmubQ"
            >
              <div
                class="sc-jTzLTM hJXymL"
              >
                <label
                  class="sc-VigVT hLPYRt"
                >
                  Risk Limit: 
                </label>
                1
                %
              </div>
              <div
                class="sc-jTzLTM hJXymL"
              >
                <label
                  class="sc-VigVT hLPYRt"
                >
                  P-value: 
                </label>
                 
                0.00020431431380638307
              </div>
            </div>
            <div
              class="sc-gqjmRU bqmubQ"
            >
              <div
                class="sc-jTzLTM hJXymL"
              >
                <label
                  class="sc-VigVT hLPYRt"
                >
                  choice one
                </label>
                <div
                  class="sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    data-testid="round-0-contest-0-choice-choice-1"
                    disabled=""
                    name="contests[0][choice-1]"
                    type="number"
                    value="100"
                  />
                </div>
              </div>
              <div
                class="sc-jTzLTM hJXymL"
              >
                <label
                  class="sc-VigVT hLPYRt"
                >
                  choice two
                </label>
                <div
                  class="sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    data-testid="round-0-contest-0-choice-choice-2"
                    disabled=""
                    name="contests[0][choice-2]"
                    type="number"
                    value="167"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Audit Progress: 
              1
               of 
              1
               
              complete
            </div>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <button
              class="sc-gzVnrw kCekyN"
            >
              Download Audit Report
            </button>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="sc-ifAKCX jXnARC"
      >
        <div
          class="sc-bxivhb doxbWk"
        >
          Round 1
        </div>
        <div
          class="sc-htpNat gYCOid"
        >
          Ballot Retrieval List: 
          379
           Total Ballots
        </div>
        <div
          class="sc-bwzfXH xctbO"
        >
          <p>
            Contest 
            1
            : 
            379
             ballots
          </p>
        </div>
        <button
          class="sc-bZQynM cjVqOQ"
        >
          Download Aggregated Ballot Retrieval List for Round 
          1
        </button>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Contest 1: contest name
          </div>
          <div
            class="sc-htpNat gYCOid"
          >
            Audited Results: Round 
            1
            , Contest 
            1
             
            COMPLETE
          </div>
          <div
            class="sc-gqjmRU bqmubQ"
          >
            <div
              class="sc-jTzLTM hJXymL"
            >
              <label
                class="sc-VigVT hLPYRt"
              >
                Risk Limit: 
              </label>
              1
              %
            </div>
            <div
              class="sc-jTzLTM hJXymL"
            >
              <label
                class="sc-VigVT hLPYRt"
              >
                P-value: 
              </label>
               
              0.00020431431380638307
            </div>
          </div>
          <div
            class="sc-gqjmRU bqmubQ"
          >
            <div
              class="sc-jTzLTM hJXymL"
            >
              <label
                class="sc-VigVT hLPYRt"
              >
                choice one
              </label>
              <div
                class="sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  data-testid="round-0-contest-0-choice-choice-1"
                  disabled=""
                  name="contests[0][choice-1]"
                  type="number"
                  value="100"
                />
              </div>
            </div>
            <div
              class="sc-jTzLTM hJXymL"
            >
              <label
                class="sc-VigVT hLPYRt"
              >
                choice two
              </label>
              <div
                class="sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  data-testid="round-0-contest-0-choice-choice-2"
                  disabled=""
                  name="contests[0][choice-2]"
                  type="number"
                  value="167"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Audit Progress: 
            1
             of 
            1
             
            complete
          </div>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <button
            class="sc-gzVnrw kCekyN"
          >
            Download Audit Report
          </button>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CalculateRiskMeasurement renders first round correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="sc-ifAKCX jXnARC"
        >
          <div
            class="sc-bxivhb doxbWk"
          >
            Round 1
          </div>
          <div
            class="sc-htpNat gYCOid"
          >
            Ballot Retrieval List: 
            379
             Total Ballots
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            <p>
              Contest 
              1
              : 
              379
               ballots
            </p>
          </div>
          <button
            class="sc-bZQynM cjVqOQ"
          >
            Download Aggregated Ballot Retrieval List for Round 
            1
          </button>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Contest 1: contest name
            </div>
            <div
              class="sc-htpNat gYCOid"
            >
              Audited Results: Round 
              1
              , Contest 
              1
               
              INCOMPLETE
            </div>
            <div
              class="sc-bwzfXH xctbO"
            >
              Enter the number of votes recorded for each candidate/choice in the audited ballots for Round 
              1
              , Contest 
              1
            </div>
            <div
              class="sc-gqjmRU bqmubQ"
            >
              <div
                class="sc-jTzLTM hJXymL"
              >
                <label
                  class="sc-VigVT hLPYRt"
                >
                  choice one
                </label>
                <div
                  class="sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    data-testid="round-0-contest-0-choice-choice-1"
                    name="contests[0][choice-1]"
                    type="number"
                    value="0"
                  />
                </div>
              </div>
              <div
                class="sc-jTzLTM hJXymL"
              >
                <label
                  class="sc-VigVT hLPYRt"
                >
                  choice two
                </label>
                <div
                  class="sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    data-testid="round-0-contest-0-choice-choice-2"
                    name="contests[0][choice-2]"
                    type="number"
                    value="0"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Audit Progress: 
              0
               of 
              1
               
              complete
            </div>
          </div>
          <div
            class="sc-gZMcBi DmELe"
          >
            <button
              class="sc-EHOje iYzqVS"
              type="button"
            >
              Calculate Risk Measurement
            </button>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          />
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="sc-ifAKCX jXnARC"
      >
        <div
          class="sc-bxivhb doxbWk"
        >
          Round 1
        </div>
        <div
          class="sc-htpNat gYCOid"
        >
          Ballot Retrieval List: 
          379
           Total Ballots
        </div>
        <div
          class="sc-bwzfXH xctbO"
        >
          <p>
            Contest 
            1
            : 
            379
             ballots
          </p>
        </div>
        <button
          class="sc-bZQynM cjVqOQ"
        >
          Download Aggregated Ballot Retrieval List for Round 
          1
        </button>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Contest 1: contest name
          </div>
          <div
            class="sc-htpNat gYCOid"
          >
            Audited Results: Round 
            1
            , Contest 
            1
             
            INCOMPLETE
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            Enter the number of votes recorded for each candidate/choice in the audited ballots for Round 
            1
            , Contest 
            1
          </div>
          <div
            class="sc-gqjmRU bqmubQ"
          >
            <div
              class="sc-jTzLTM hJXymL"
            >
              <label
                class="sc-VigVT hLPYRt"
              >
                choice one
              </label>
              <div
                class="sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  data-testid="round-0-contest-0-choice-choice-1"
                  name="contests[0][choice-1]"
                  type="number"
                  value="0"
                />
              </div>
            </div>
            <div
              class="sc-jTzLTM hJXymL"
            >
              <label
                class="sc-VigVT hLPYRt"
              >
                choice two
              </label>
              <div
                class="sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  data-testid="round-0-contest-0-choice-choice-2"
                  name="contests[0][choice-2]"
                  type="number"
                  value="0"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Audit Progress: 
            0
             of 
            1
             
            complete
          </div>
        </div>
        <div
          class="sc-gZMcBi DmELe"
        >
          <button
            class="sc-EHOje iYzqVS"
            type="button"
          >
            Calculate Risk Measurement
          </button>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        />
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
